flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+//===-- CallInterface.cpp -- Procedure call interface ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Lower/CallInterface.h"
+#include "flang/Evaluate/characteristics.h"
+#include "flang/Evaluate/fold.h"
+#include "flang/Lower/Bridge.h"
+#include "flang/Lower/FIRBuilder.h"
+#include "flang/Lower/PFTBuilder.h"
+#include "flang/Optimizer/Dialect/FIRDialect.h"
+#include "flang/Optimizer/Support/InternalNames.h"
+#include "flang/Semantics/symbol.h"
+#include "flang/Semantics/tools.h"
+
+//===----------------------------------------------------------------------===//
+// Caller side interface implementation
+//===----------------------------------------------------------------------===//
+
+bool Fortran::lower::CallerInterface::hasAlternateReturns() const {
+  return procRef.HasAlternateReturns();
+}
+
+std::string Fortran::lower::CallerInterface::getMangledName() const {
+  const auto &proc = procRef.proc();
+  if (const auto *symbol = proc.GetSymbol())
+    return converter.mangleName(*symbol);
+  assert(proc.GetSpecificIntrinsic() &&
+         "expected intrinsic procedure in designator");
+  return proc.GetName();
+}
+
+mlir::Location Fortran::lower::CallerInterface::getCalleeLocation() const {
+  const auto &proc = procRef.proc();
+  if (const auto *symbol = proc.GetSymbol()) {
+    // FIXME: If the callee is defined in the same file but after the current
+    // unit we do cannot get its location here and the funcOp is created at the
+    // wrong location (i.e, the caller location).
```

Comment:
I assumed that the front-end was going to effectively do all this resolution for the bridge and it would be a matter of recovering that information from the symbol table. That may not be a good assumption.

Reply:
Let's make that a requirement if that's the right thing to do.   Please check with @tskeith.

Reply:
Good point, that would be better. That could go with the analysis of F77 call sites to check they are compatible (that we discussed the other day), it could be a good time to place the symbol of the actual procedure definition if it exists in this file.
Going through all the funit from the PFT to generate their mlir::FuncOp first is probably also more robust. We could store CallerInterface in the funit to access it later when lowering the whole funit.

Also, this discussion is not limited to F77, one can very well write:
```
subroutine foo
  interface
   subroutine bar(x, y)
      real, allocatable :: x(:, :), y(:)
   end subroutine bar
  end interface
  allocatable :: x(:, :), y(:)
  call bar(x, y)
end subroutine

subroutine bar(z)
 character, allocatable :: z(:)
end subroutine

call foo
end
```
That's also very bad (bar interface does not match the actual definition in the same file). gfortran emits a warning but compiles and link, and pgfortran compiles and link silently. There is no way the binary is correct. We would crash during lowering since there can only be one external procedure "bar". It would be nice if f18 semantics could fire an nice error before.

I have added a comment to the issue already tracking the F77 interface issue.

Reply:
I agree that it would be much nicer to issue a diagnostic here. The user should, by default, be given a heads-up that something tragic is about to happen in their executable. :)

And yes, I suggested a separate pass over the interfaces for this exact issue. Even when an interface is provided, it can be incorrect.